bootpef.8
.\" bootpef.8
.TH BOOTPEF 8 "4 Dec 1993" "MAINTENANCE COMMANDS"
.SH NAME
bootpef \- BOOTP Extension File compiler
.SH SYNOPSIS
.LP
.B bootpef
.RB [ " \-c"
.IR chdir " ]"
.RB [ " \-d"
.IR debug-level " ]"
.RB [ "-f"
.IR config-file " ]"
.RI "[ " client-name " [...]]"
.SH DESCRIPTION
.B bootpef
builds the
.I Extension Path
files described by RFC 1497 (tag 18).
If any
.I client-name
arguments are specified, then
.I bootpef
compiles the extension files for only those clients.
.SH OPTIONS
.TP
.BI \-c \ chdir\-path
Sets the current directory used by
.I bootpef
while creating extension files.  This is useful when the
extension file names are specified as relative pathnames, and
.I bootpef
needs to use the same current directory as the TFTP server
(typically 
.IR /tftpboot ).
.TP
.BI \-d \ debug\-level
Sets the
.I debug\-level
variable that controls the amount of debugging messages generated.
For example, -d4 or -d 4 will set the debugging level to 4.
.TP
.BI \-f \ config\-file
Set the name of the config file that specifies the option
data to be sent to each client.
.SH "SEE ALSO"
.BR bootpd (8), 
.BR tftpd (8)
.SH REFERENCES
.TP
RFC951
BOOTSTRAP PROTOCOL (BOOTP)
.TP
RFC1497
BOOTP Vendor Information Extensions




Changes
Changes, most recent first
Date, <email> Real Name
	what...

--> bootp-2.4.3

03/27/96  gwr@mc.com (Gordon W. Ross)
	Use LOG_NOTICE in place of LOG_INFO for messages related
	to unsatisfied clients [at request of <otto@tukki.jyu.fi>]
	Fix the irix Makefile targets, and other misc.

03/25/95  gwr@mc.com (Gordon W. Ross)
	Corrected a bug I introduced into SunOS setarp, where
	bad IP address caused "network unreachable" errors.
	[Thanks to andrew@ntplx.net (Andrew Lindh) for the fix!]

--> bootpd-2.4.3

11/19/95  joey@infodrom.north.de (Martin Schulze)
	Fixed a problem that DHCP lease time is send in every request
	independend of the existence of a dl tag.
	Corrected some NEED()-calls with wrong arguments (only some text)
	Added support for -v (version) in bootpd and bootpgw.

08/07/95  middelin@polyware.iaf.nl (Pauline Middelink)
	Fixed problem while reporting duplicate hardware address in
	readfile.c (kerry@maine.maine.edu)
	Changed some code to allow the giaddr to pass through. Needed
	when the client request is forwarded to us through a gateway.
	(Broadcasting to the gateway will not reach the client)
	
01/26/95  Phil.Packer@bbc.co.uk (Phil.Packer)
	Added dl tag for DHCP lease and rejigged Pauline's codes
	at the end of bootpd.c to allow the setting of explicit leases.
	Added "sco" Makefile target.
	
--> bootp-2.4.2

01/14/95  middelin@polyware.iaf.nl (Pauline Middelink)
	Corrected support for the Linux networking code.
	Fixed lots of warnings (gcc -Wall)
	Added "linux" Makefile target.
	Written code for DHCP static IP allocation.

01/02/95  Jukka Ukkonen <ukkonen@csc.fi>
	Allow bootptab syntax: ha="0:0:c0:80:e8:a7"

11/30/94  Tonny van Lankveld <A.L.M.G.v.Lankveld@urc.tue.nl>
	Fix reporting of duplicate Ethernet addresses.

09/06/94  longyear@netcom.com (Al Longyear)
	Better setarp for linux, allows non-ether types.

09/02/94  Robert MacKinnon <rbm@montrouge.mis.slb.com>
	Add support for IBM's AIX 3.2.5

08/30/94  piercarl@ltd.c-d.com (Piercarlo Grandi)
	Fix select calls on linux (modifies timeval arg).
	Fix setarp (specify Ethernet type for now).

08/27/94  drew@drewsun.FEITH.COM (Andrew B. Sudell)
	Add support for Wollongong Win-TCP (SysVr4 variant).

08/24/94  gwr@mc.com (Gordon W. Ross)
	Use sigaction() on systems that define SA_NOCLDSTOP
	(a symbol required by POSIX) for HP/UX and others.

--> bootp-2.4.1

08/24/94  gwr@mc.com (Gordon W. Ross)
	Fix bug in boot file name generation (missing init)

--> bootp-2.4.0

08/20/94  gwr@mc.com (Gordon W. Ross)
	Fix code to build bootfile name based on combination of
	client requested name and bootfile specifications.
	Behave similarly with or without CHECK_FILE_ACCESS.

07/30/94  Dirk Koeppen <dirk@incom.de>
	Add "min wait" option (mw) to cause bootpd to ignore
	requests from clients that have not waited long enough.
	Add code to honor client requests containing the DHCP
	option "Maximum Message Size" and use its value to
	determine the size of the reply message.

--> bootp-2.3.8

06/25/94  Christos Zoulas <christos@deshaw.com>
	Add "-h" flag to override host name (affects default IP
	address provided in reply messages.  (Also minor bug fix)

05/27/94  gwr@mc.com (Gordon W. Ross)
	Add code to call "arp -s IPADDR HWADDR" on systems
	that do not provide an SIOCSARP ioctl (i.e. NetBSD)

--> bootp-2.3.7

05/05/94  Walter Wong <wcw+@CMU.EDU>
	Reduce noize at debug level one, where log messages
	are generated only for hosts that are recognized
	and replied to by bootpd.  (At request of HP folks.)

04/30/94  gwr@mc.com (Gordon W. Ross)
	Use memxxx functions unless USE_BFUNCS is defined.
	Added -f <file> option to bootptest (requested file).

04/29/94  tpaquett@ita.lgc.com (Trevor Paquette)
	Remove call to haddr_conv802() in sendreply().
	The setarp should get the non-transformed address.

04/27/94  gwr@mc.com
	Improve logic for building bootfile pathname, so a path
	will be put in the reply if either the client or bootpd
	specifies a boot file.  (Needed for NetBSD diskless boot)

04/25/94  shamash@boxhill.com (Ari Shamash)
	Fix prs_inetaddr() so it allows '_' in hostnames.

04/16/94  gwr@mc.com (Gordon W. Ross)
	Fix setarp for SVR4 (needs to use I_STR ioctl)
	Thanks to several people: (all sent the same fix)
		Barney Wolff <barney@databus.com>,
		bear@upsys.se (Bj|rn Sj|holm),
		Michael Kuschke <Michael.Kuschke@Materna.DE>,

03/25/95  Ulrich Heuer </I=zhhi9/G=Ulrich/S=Heuer/@zhflur.ubs.ubs.ch>
	Make option string lengths not include a null terminator.
	The trailing null breaks some clients.

03/15/94  "Edmund J. Sutcliffe" <ejs1@tower.york.ac.uk>
	Add support for the "EX" option:  Execute a program
	before sending a BOOTREPLY to a client.  Support for
	this option is conditional on YORK_EX_OPTION.

03/10/94  Nigel Metheringham <nigelm@ohm.york.ac.uk>
	Make getether.c work on Linux.

03/09/94  Koch@Math.Uni-Duisburg.DE (Peter Koch)
	Add missing MANDIR definition to Makefile.

03/08/94  Jeroen.Scheerder@let.ruu.nl
	Fix args to report in getether code for Ultrix.
	Run install individually for each program.

--> bootp-2.3.6
03/07/94  gwr@mc.com
	Cleanup for release (run gnu indent, tab-size=4)

02/24/94  Jeroen.Scheerder@let.ruu.nl
	Allow underscore in host names - readfile.c:goodname()
	Add ConvOldTab.sh - converts 1.1 bootptab to new format.

02/20/94  gwr@mc.com (Gordon W. Ross)
	Make readfile tolerant of hardware addresses that start
	with a letter.  (If lookup_hwa() fails, assume numeric.)
	Fix whitespace skip before :vm= auto: and avoid lookup.

02/12/94  walker@zk3.dec.com (Mary Walker)
	Added support for 64-bit longs (for the DEC Alpha)
	Allow ieee802 hardware address in bit-reversed oreder

02/07/94  hl@tekla.fi (Harald Lundberg)
	Fix conflict with DUMP_FILE in syslog.h on OSF1
	Use int for (struct bootp).bp_xid (for DEC Alpha)
	Added Ultrix support to bootptest (getether)

02/06/94  brezak@ch.hp.com (John Brezak)
	Add man-page and install targets to Makefile.NetBSD
	Add getether support for NetBSD

02/05/94  gwr@mc.com (Gordon W. Ross)

bootpef.c
/************************************************************************
          Copyright 1988, 1991 by Carnegie Mellon University

                          All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted, provided
that the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation, and that the name of Carnegie Mellon University not be used
in advertising or publicity pertaining to distribution of the software
without specific, written prior permission.

CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS
SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.
IN NO EVENT SHALL CMU BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.
************************************************************************/

/*
 * bootpef - BOOTP Extension File generator
 *	Makes an "Extension File" for each host entry that
 *	defines an and Extension File. (See RFC1497, tag 18.)
 *
 * HISTORY
 *	See ./Changes
 *
 * BUGS
 *	See ./ToDo
 */



#ifdef	__STDC__
#include <stdarg.h>
#else
#include <varargs.h>
#endif

#include <sys/types.h>
#include <sys/time.h>

#include <netinet/in.h>
#include <arpa/inet.h>			/* inet_ntoa */

#ifndef	NO_UNISTD
#include <unistd.h>
#endif
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <syslog.h>

#ifndef	USE_BFUNCS
#include <memory.h>
/* Yes, memcpy is OK here (no overlapped copies). */
#define bcopy(a,b,c)    memcpy(b,a,c)
#define bzero(p,l)      memset(p,0,l)
#define bcmp(a,b,c)     memcmp(a,b,c)
#endif

#include "bootp.h"
#include "hash.h"
#include "hwaddr.h"
#include "bootpd.h"
#include "dovend.h"
#include "readfile.h"
#include "report.h"
#include "tzone.h"
#include "patchlevel.h"

#define	BUFFERSIZE   		0x4000

#ifndef CONFIG_FILE
#define CONFIG_FILE		"/etc/bootptab"
#endif



/*
 * Externals, forward declarations, and global variables
 */

#ifdef	__STDC__
#define P(args) args
#else
#define P(args) ()
#endif

static void mktagfile P((struct host *));
static void usage P((void));

#undef P


/*
 * General
 */

char *progname;
char *chdir_path;
int debug = 0;					/* Debugging flag (level) */
byte *buffer;

/*
 * Globals below are associated with the bootp database file (bootptab).
 */

char *bootptab = CONFIG_FILE;


/*
 * Print "usage" message and exit
 */
static void
usage()
{
	fprintf(stderr,
	   "usage:  $s [ -c chdir ] [-d level] [-f configfile] [host...]\n");
	fprintf(stderr, "\t -c n\tset current directory\n");
	fprintf(stderr, "\t -d n\tset debug level\n");
	fprintf(stderr, "\t -f n\tconfig file name\n");
	exit(1);
}


/*
 * Initialization such as command-line processing is done and then the
 * main server loop is started.
 */
int
main(argc, argv)
	int argc;
	char **argv;
{
	struct host *hp;
	char *stmp;
	int n;

	progname = strrchr(argv[0], '/');
	if (progname) progname++;
	else progname = argv[0];

	/* Get work space for making tag 18 files. */
	buffer = (byte *) malloc(BUFFERSIZE);
	if (!buffer) {
		report(LOG_ERR, "malloc failed");
		exit(1);
	}
	/*
	 * Set defaults that might be changed by option switches.
	 */
	stmp = NULL;

	/*
	 * Read switches.
	 */
	for (argc--, argv++; argc > 0; argc--, argv++) {
		if (argv[0][0] != '-')
			break;
		switch (argv[0][1]) {

		case 'c':				/* chdir_path */
			if (argv[0][2]) {
				stmp = &(argv[0][2]);
			} else {
				argc--;
				argv++;
				stmp = argv[0];
			}
			if (!stmp || (stmp[0] != '/')) {
				fprintf(stderr,
						"bootpd: invalid chdir specification\n");
				break;
			}
			chdir_path = stmp;
			break;

		case 'd':				/* debug */
			if (argv[0][2]) {
				stmp = &(argv[0][2]);
			} else if (argv[1] && argv[1][0] == '-') {
				/*
				 * Backwards-compatible behavior:
				 * no parameter, so just increment the debug flag.
				 */
				debug++;
				break;
			} else {
				argc--;
				argv++;
				stmp = argv[0];
			}
			if (!stmp || (sscanf(stmp, "%d", &n) != 1) || (n < 0)) {
				fprintf(stderr,
						"bootpd: invalid debug level\n");
				break;
			}
			debug = n;
			break;

		case 'f':				/* config file */
			if (argv[0][2]) {
				stmp = &(argv[0][2]);
			} else {
				argc--;
				argv++;
				stmp = argv[0];
			}
			bootptab = stmp;
			break;

		default:
			fprintf(stderr, "bootpd: unknown switch: -%c\n",
					argv[0][1]);
			usage();
			break;
		}
	}

	/* Get the timezone. */
	tzone_init();

	/* Allocate hash tables. */
	rdtab_init();

	/*
	 * Read the bootptab file.
	 */
	readtab(1);					/* force read */

	/* Set the cwd (i.e. to /tftpboot) */
	if (chdir_path) {
		if (chdir(chdir_path) < 0)
			report(LOG_ERR, "%s: chdir failed", chdir_path);
	}
	/* If there are host names on the command line, do only those. */
	if (argc > 0) {
		unsigned int tlen, hashcode;

		while (argc) {
			tlen = strlen(argv[0]);
			hashcode = hash_HashFunction((u_char *)argv[0], tlen);
			hp = (struct host *) hash_Lookup(nmhashtable,
											 hashcode,
											 nmcmp, argv[0]);
			if (!hp) {
				printf("%s: no matching entry\n", argv[0]);
				exit(1);
			}
			if (!hp->flags.exten_file) {
				printf("%s: no extension file\n", argv[0]);
				exit(1);
			}
			mktagfile(hp);
			argv++;
			argc--;
		}
		exit(0);
	}
	/* No host names specified.  Do them all. */
	hp = (struct host *) hash_FirstEntry(nmhashtable);
	while (hp != NULL) {
		mktagfile(hp);
		hp = (struct host *) hash_NextEntry(nmhashtable);
	}

	return 0;
}



/*
 * Make a "TAG 18" file for this host.
 * (Insert the RFC1497 options.)
 */

static void
mktagfile(hp)
	struct host *hp;
{
	FILE *fp;
	int bytesleft, len;
	byte *vp;

	if (!hp->flags.exten_file)
		return;

	vp = buffer;
	bytesleft = BUFFERSIZE;
	bcopy(vm_rfc1048, vp, 4);	/* Copy in the magic cookie */
	vp += 4;
	bytesleft -= 4;

	/*
	 * The "extension file" options are appended by the following
	 * function (which is shared with bootpd.c).
	 */
	len = dovend_rfc1497(hp, vp, bytesleft);
	vp += len;
	bytesleft -= len;

	if (bytesleft < 1) {
		report(LOG_ERR, "%s: too much option data",
			   hp->exten_file->string);
		return;
	}
	*vp++ = TAG_END;
	bytesleft--;

	/* Write the buffer to the extension file. */
	printf("Updating \"%s\"\n", hp->exten_file->string);
	if ((fp = fopen(hp->exten_file->string, "w")) == NULL) {
		report(LOG_ERR, "error opening \"%s\": %s",
			   hp->exten_file->string, get_errmsg());
		return;
	}
	len = vp - buffer;
	if (len != fwrite(buffer, 1, len, fp)) {
		report(LOG_ERR, "write failed on \"%s\" : %s",
			   hp->exten_file->string, get_errmsg());
	}
	fclose(fp);

} /* mktagfile */

/*
 * Local Variables:
 * tab-width: 4
 * c-indent-level: 4
 * c-argdecl-indent: 4
 * c-continued-statement-offset: 4
 * c-continued-brace-offset: -4
 * c-label-offset: -4
 * c-brace-offset: 0
 * End:
 */
